lib/puppet/pops/lookup/data_provider.rb

```diff
@@ -0,0 +1,84 @@
+module Puppet::Pops
+module Lookup
+# @api private
+module DataProvider
+  # The Pcore type for all keys and subkeys in a data hash.
+  TYPE_DATA_KEY = Types::PVariantType.new([
+    Types::PStringType::NON_EMPTY,
+    Types::PBooleanType::DEFAULT,
+    Types::PNumericType::DEFAULT
+  ])
+
+  # The Pcore type for all values and sub-values in a data hash.
+  TYPE_DATA_VALUE = Types::PVariantType.new([
+    Types::PScalarType::DEFAULT,
+    Types::PUndefType::DEFAULT,
+    Types::PHashType::DEFAULT,
+    Types::PArrayType::DEFAULT
+  ])
```

Comment:
As noted elsewhere, any Pcore representable type except RuntimeType should be allowed as a value.